### PR TITLE
feat(tree): enable add at location

### LINKED ISSUE #4049: Add index to Tree widget's add
Presently there is no way to add a new tree node except as the last node, accordingly:

Would you accept an optional `index`here?
https://github.com/Textualize/textual/blob/ce962196c73b2b87c9f812f66a6f6ded3f2aaf4b/src/textual/widgets/_tree.py#L320-L327

If set it would change this from `append` to `insert` and pass `index`:
https://github.com/Textualize/textual/blob/ce962196c73b2b87c9f812f66a6f6ded3f2aaf4b/src/textual/widgets/_tree.py#L344

### PR DESCRIPTION
**Pull request by TomJGooding**
Add `before` and `after` parameters to the `TreeNode.add` and `TreeNode.add_leaf` methods to enable positioning a new node.

Closes #4049.

**Please review the following checklist.**

- [x] Docstrings on all new or modified functions / classes 
- [ ] Updated documentation
- [x] Updated CHANGELOG.md (where appropriate)

### CODE REVIEW

File: src/textual/widgets/_tree.py [hunk 1]
```diff
@@ -330,34 +336,101 @@ def add(
...
+
+        if before is not None:
+            if isinstance(before, int):
+                insert_index = before
+            elif isinstance(before, TreeNode):
+                try:
+                    insert_index = self.children.index(before)
+                except ValueError:
+                    raise AddNodeError(
+                        "Request to add before an unknown node that is not a child of this node"
```
Comment (darrenburns) [inline src/textual/widgets/_tree.py:367]: Struggled to mentally parse the error message a bit here.

```suggestion
                        "The node specified for 'before' does not exist in this Tree."
```

File: src/textual/widgets/_tree.py [hunk 2]
```diff
@@ -330,34 +336,101 @@ def add(
...
+            if isinstance(after, int):
+                insert_index = after + 1
+                if after < 0:
+                    insert_index += len(self.children)
+            elif isinstance(after, TreeNode):
+                try:
+                    insert_index = self.children.index(after) + 1
+                except ValueError:
+                    raise AddNodeError(
+                        "Request to add after an unknown node that is not a child of this node"
```
Comment (darrenburns) [inline src/textual/widgets/_tree.py:384]: Same as above.

**Review by darrenburns — CHANGES_REQUESTED** [review]
Minor request to change the wording of an error message, but otherwise looks good to me! Will be very useful - thanks!

File: src/textual/widgets/_tree.py (hunk 1 above)
Comment (darrenburns) [reply src/textual/widgets/_tree.py:367]: Actually on reading the code in more detail I think my suggestion isn't clear.

Maybe something like `f"cannot insert before {before}, as this node only has {len(children)} children"`?

Comment (darrenburns) [reply src/textual/widgets/_tree.py:367]: Haha yeah, I realised that after posting. See my other comment 😄

Comment (TomJGooding) [reply src/textual/widgets/_tree.py:367]: > f"cannot insert before {before}, as this node only has {len(children)} children"

Sorry but I'm not sure that's quite right either. The length of the children isn't relevant, rather that the `TreeNode` passed as the `before` isn't a child node of this node?

Comment (darrenburns) [reply src/textual/widgets/_tree.py:367]: ah sorry, i thought we were in the `isinstance(before, int)` branch 🤦 

how about:

```suggestion
                        "The node specified for `before` is not a child of this node"
```

Comment (TomJGooding) [reply src/textual/widgets/_tree.py:367]: Sorry also just to clarify, if the `before`/`after` argument is an index that is out of range, this also won't raise an error (the same as Python's `insert`)

Comment (darrenburns) [reply src/textual/widgets/_tree.py:367]: Ah ok - that seems fine to me (I wasn't aware that's how insert worked)

**Comment by willmcgugan** [comment]
Merged the CHANGELOG.

Thanks, Tom. I think we're going to squeeze this in to 0.73.0, which I'll release at EOD.

File: src/textual/widgets/_tree.py (hunk 1 above)
Comment (TomJGooding) [reply src/textual/widgets/_tree.py]: Thanks, that wording is much better! Updated in 04ac1ca (sorry I tried adding you as co-author but always forget that second hyphen!)

**Review by darrenburns — APPROVED** [review]

**Pull request merged**